cmd/kubepkg/cmd/root.go

```diff
@@ -63,109 +47,114 @@ func Execute() {
-		&rootOpts.cniVersion,
+		&cniVersion,
 		"cni-version",
-		"",
+		opts.CNIVersion(),
 		"CNI version to build",
 	)
+
+	criToolsVersion := ""
 	rootCmd.PersistentFlags().StringVar(
-		&rootOpts.criToolsVersion,
+		&criToolsVersion,
 		"cri-tools-version",
-		"",
+		opts.CRIToolsVersion(),
 		"CRI tools version to build",
 	)
+
+	releaseDownloadLinkBase := ""
 	rootCmd.PersistentFlags().StringVar(
-		&rootOpts.releaseDownloadLinkBase,
+		&releaseDownloadLinkBase,
 		"release-download-link-base",
-		kpkg.DefaultReleaseDownloadLinkBase,
+		opts.ReleaseDownloadLinkBase(),
 		"release download link base",
 	)
+
+	templateDir := ""
 	rootCmd.PersistentFlags().StringVar(
-		&rootOpts.templateDir,
+		&templateDir,
 		"template-dir",
-		kpkg.LatestTemplateDir,
+		opts.TemplateDir(),
 		"template directory",
 	)
+
+	specOnly := false
 	rootCmd.PersistentFlags().BoolVar(
-		&rootOpts.specOnly,
+		&specOnly,
 		"spec-only",
-		false,
+		opts.SpecOnly(),
 		"only create specs instead of building packages",
 	)
+
 	rootCmd.PersistentFlags().StringVar(
-		&rootOpts.logLevel,
+		&logLevel,
 		"log-level",
 		"info",
 		"the logging verbosity, either 'panic', 'fatal', 'error', 'warn', 'warning', 'info', 'debug' or 'trace'",
 	)
-}
 
-// initConfig reads in config file and ENV variables if set.
-func initConfig() {
+	opts = opts.WithPackages(packages...).
+		WithChannels(channels...).
+		WithArchitectures(architectures...).
+		WithKubeVersion(kubeVersion).
+		WithRevision(revision).
+		WithCNIVersion(cniVersion).
+		WithCRIToolsVersion(criToolsVersion).
+		WithReleaseDownloadLinkBase(releaseDownloadLinkBase).
+		WithTemplateDir(templateDir).
+		WithSpecOly(specOnly)
 }
 
 func initLogging(*cobra.Command, []string) error {
-	return log.SetupGlobalLogger(rootOpts.logLevel)
+	return log.SetupGlobalLogger(logLevel)
 }
 
-func (r *rootOptions) validate() error {
-	if ok := kpkg.IsSupported(r.packages, kpkg.SupportedPackages); !ok {
-		return errors.New("package selections are not supported")
-	}
-	if ok := kpkg.IsSupported(r.channels, kpkg.SupportedChannels); !ok {
-		return errors.New("channel selections are not supported")
-	}
-	if ok := kpkg.IsSupported(r.architectures, kpkg.SupportedArchitectures); !ok {
-		return errors.New("architectures selections are not supported")
-	}
-
-	// Replace the "+" with a "-" to make it semver-compliant
-	r.kubeVersion = util.TrimTagPrefix(r.kubeVersion)
-
-	return nil
-}
-
-func run(ro *rootOptions, buildType kubepkg.BuildType) error {
-	client := kubepkg.New()
-	builds, err := client.ConstructBuilds(
-		buildType, ro.packages, ro.channels, ro.kubeVersion, ro.revision,
-		ro.cniVersion, ro.criToolsVersion, ro.templateDir,
-	)
+func run(opts *options.Options) error {
+	client := kubepkg.New(opts)
+	builds, err := client.ConstructBuilds()
 	if err != nil {
 		return errors.Wrap(err, "running kubepkg")
 	}
-	return client.WalkBuilds(builds, ro.architectures, ro.specOnly)
+	return client.WalkBuilds(builds)
 }
```

Comment:
I think this looks like we wanted to have it. Ref https://github.com/kubernetes/release/pull/1282/files#r424507497